ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/BlockchainQueries.java

```diff
@@ -154,6 +154,22 @@ public long headBlockNumber() {
         address, blockNumber, account -> account.getStorageValue(storageIndex), UInt256.ZERO);
   }
 
+  /**
+   * Determines the block header for the address associated with this storage index.
+   *
+   * @param address The address of the account that owns the storage being queried.
+   * @param storageIndex The storage index whose value is being retrieved.
+   * @param blockHash The blockHash that is being queried.
+   * @return The value at the storage index being queried.
+   */
+  public Optional<UInt256> storageAt(
+      final Address address, final UInt256 storageIndex, final Hash blockHash) {
+    final long blockNumber =
+        getBlockHeaderByHash(blockHash).map(BlockHeader::getNumber).orElse(Long.MAX_VALUE);
+
+    return storageAt(address, storageIndex, blockNumber);
```

Comment:
We can't simply do a hash->BlockNumber query here.  What if they are querying from an orphaned block hash?  The numbers will be wrong.

`storageAt` and the whole call series needs to be primarily hash based, where the calls to get it by block number quickly go to the hash based call chain.